packages/abi-coder/src/utils/getFunctionInputs.ts

```diff
@@ -0,0 +1,33 @@
+import type { JsonAbi, JsonAbiArgument } from '../types/JsonAbi';
+
+import { OPTION_REGEX, VOID_TYPE } from './constants';
+import { findTypeById } from './json-abi';
+
+export type ArgumentWithMetadata<TArg extends JsonAbiArgument = JsonAbiArgument> = TArg & {
+  isOptional: boolean;
+};
+
+export const getFunctionInputs = (params: {
```

Comment:
A few optimization suggestions I've noticed: 

- Deep copies can be quite expensive, especially on more complex objects, better to work with the original inputs
- reverses the array twice (.reverse() at the start and end).
-  creates multiple intermediate arrays (`reverse()`, `map` and `reverse()` again)

```suggestion
export const getFunctionInputs = ({
  jsonAbi,
  inputs,
}: {
  jsonAbi: JsonAbi;
  inputs: readonly JsonAbiArgument[];
}): ArgumentWithMetadata[] => {
  let isMandatory = false;

  return inputs.reduceRight((result, input) => {
    const type = findTypeById(jsonAbi, input.type);
    isMandatory = isMandatory || (type.type !== VOID_TYPE && !OPTION_REGEX.test(type.type));

    return [{ ...input, isOptional: !isMandatory }, ...result];
  }, [] as ArgumentWithMetadata[]);
 ```

Reply:
Thank you for the optimizations 🙏🏼 

71a1309b87050e391251fef84fe2cd57ded1afe6